// payjp/built/index.js
"use strict";
const accounts_1 = require("./accounts");
const balances_1 = require("./balances");
const charges_1 = require("./charges");
const customers_1 = require("./customers");
const events_1 = require("./events");
const plans_1 = require("./plans");
const statements_1 = require("./statements");
const subscriptions_1 = require("./subscriptions");
const tenants_1 = require("./tenants");
const tenantTransfers_1 = require("./tenantTransfers");
const terms_1 = require("./terms");
const threeDSecureRequests_1 = require("./threeDSecureRequests");
const tokens_1 = require("./tokens");
const transfers_1 = require("./transfers");
const Payjp = function (apikey, options = {}) {
    if (!apikey) {
        throw new Error("Please set apikey.");
    }
    else if (apikey.indexOf("pk_") === 0) {
        throw new Error("You cannot use the public apikey in this SDK.");
    }
    const payjpConfig = {
        apikey,
        config: {
            apibase: options.apibase || "https://api.pay.jp/v1",
            timeout: options.timeout || 0,
            maxRetry: options.maxRetry || 0,
            retryInitialDelay: options.retryInitialDelay || 2000,
            retryMaxDelay: options.retryMaxDelay || 32000,
        },
    };
    return {
        ...payjpConfig,
        charges: new charges_1.default(payjpConfig),
        customers: new customers_1.default(payjpConfig),
        plans: new plans_1.default(payjpConfig),
        subscriptions: new subscriptions_1.default(payjpConfig),
        tokens: new tokens_1.default(payjpConfig),
        transfers: new transfers_1.default(payjpConfig),
        events: new events_1.default(payjpConfig),
        accounts: new accounts_1.default(payjpConfig),
        tenants: new tenants_1.default(payjpConfig),
        tenant_transfers: new tenantTransfers_1.default(payjpConfig),
        statements: new statements_1.default(payjpConfig),
        terms: new terms_1.default(payjpConfig),
        balances: new balances_1.default(payjpConfig),
        three_d_secure_requests: new threeDSecureRequests_1.default(payjpConfig),
    };
};
module.exports = Payjp;
