// payjp/built/resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Resource {
    payjp;
    constructor(payjp) {
        this.payjp = payjp;
    }
    get apibase() {
        return this.payjp.config.apibase;
    }
    get apikey() {
        return this.payjp.apikey;
    }
    buildHeader(method) {
        const encodedKey = Buffer.from(`${this.payjp.apikey}:`).toString("base64");
        const headers = {
            Accept: "application/json",
            Authorization: `Basic ${encodedKey}`,
        };
        if (method === "POST" || method === "PUT") {
            headers["Content-Type"] = "application/x-www-form-urlencoded";
        }
        return headers;
    }
    getCurrentDelay(retryCount, initialDelay, maxDelay) {
        const delay = Math.min(initialDelay * 2 ** retryCount, maxDelay);
        return Math.ceil((delay / 2) * (1 + Math.random()));
    }
    isPlainObject(value) {
        return Object.prototype.toString.call(value) === "[object Object]";
    }
    appendFormValue(params, key, value) {
        if (value === undefined) {
            return;
        }
        if (this.isPlainObject(value)) {
            for (const [nestedKey, nestedValue] of Object.entries(value)) {
                this.appendFormValue(params, `${key}[${nestedKey}]`, nestedValue);
            }
            return;
        }
        params.append(key, String(value));
    }
    serializeQuery(query) {
        const params = new URLSearchParams();
        for (const [key, value] of Object.entries(query)) {
            this.appendFormValue(params, key, value);
        }
        return params;
    }
    request(method, endpoint, query = {}, headers = {}) {
        const header = Object.assign(this.buildHeader(method), headers);
        const doRequest = async () => {
            let url = `${this.payjp.config.apibase}/${endpoint}`;
            const fetchOptions = {
                method,
                headers: header,
            };
            // Set query parameters or request body
            if (method === "GET" || method === "DELETE") {
                // For GET and DELETE, add query parameters to URL
                const params = this.serializeQuery(query);
                const queryString = params.toString();
                if (queryString) {
                    url = `${url}?${queryString}`;
                }
            }
            else {
                // For POST and PUT, send as request body
                const body = this.serializeQuery(query);
                fetchOptions.body = body.toString();
            }
            // Set timeout
            if (this.payjp.config.timeout > 0) {
                const controller = new AbortController();
                fetchOptions.signal = controller.signal;
                setTimeout(() => controller.abort(), this.payjp.config.timeout);
            }
            try {
                const response = await fetch(url, fetchOptions);
                // fetch doesn't reject on HTTP error status, so check manually
                if (!response.ok) {
                    // Parse and return error response
                    const errorBody = await response.json().catch(() => ({}));
                    const error = new Error(`HTTP Error ${response.status}`);
                    error.status = response.status;
                    error.body = errorBody;
                    error.message = errorBody.message || `HTTP Error ${response.status}`;
                    // Set response.body for compatibility with superagent
                    error.response = { body: errorBody, status: response.status };
                    throw error;
                }
                return response;
            }
            catch (error) {
                // Convert timeout error message
                if (error.name === "AbortError" || error.name === "TimeoutError") {
                    const timeoutError = new Error(`Timeout of ${this.payjp.config.timeout}ms exceeded`);
                    timeoutError.name = "TimeoutError";
                    timeoutError.code = "ETIMEDOUT";
                    timeoutError.timeout = this.payjp.config.timeout;
                    throw timeoutError;
                }
                // Convert network error message
                if (error.cause) {
                    if (error.cause.code === "ECONNRESET") {
                        const networkError = new Error("socket hang up");
                        networkError.code = "ECONNRESET";
                        throw networkError;
                    }
                }
                // Convert 'fetch failed' error to 'socket hang up' (when connection is terminated)
                if (error.message?.includes("fetch failed")) {
                    const networkError = new Error("socket hang up");
                    networkError.code = "ECONNRESET";
                    throw networkError;
                }
                throw error;
            }
        };
        let retryCount = 0;
        const retry = (resolve, reject) => doRequest()
            .then(async (res) => {
            // Try to parse as JSON. If it fails, return empty object
            try {
                const body = await res.json();
                resolve(body);
            }
            catch (_e) {
                // For non-JSON responses (HTML, etc.), return empty object
                resolve({});
            }
        })
            .catch((error) => {
            // Retry for 429 (Too Many Requests)
            if (error.status === 429 && retryCount < this.payjp.config.maxRetry) {
                const delayWithJitter = this.getCurrentDelay(retryCount, this.payjp.config.retryInitialDelay, this.payjp.config.retryMaxDelay);
                retryCount++;
                setTimeout(() => retry(resolve, reject), delayWithJitter);
            }
            else {
                return reject(error);
            }
        });
        return new Promise(retry);
    }
}
exports.default = Resource;
